File: examples/addressbook/addressbook/impl/resolver.py
# coding: utf-8
"""
The one piece of validation you write by hand.

Firestone works out *what* every rule needs; this works out *how* to fetch it.
Swap the dictionaries below for your database, ORM or cache and nothing else
about the generated validation package changes.

Every lookup a single request needs arrives in one call, already de-duplicated,
so a request that trips ten rules still costs one round trip.
"""

import collections
import logging

from fastapi.responses import JSONResponse

from addressbook.validation import RefRequest
from addressbook.validation import ValidationError

_LOGGER = logging.getLogger(__name__)

# Stand-ins for the tables a real service would query.
PERSONS = {
    "foo": {"first_name": "foo", "last_name": "bar", "uuid": "3fa8-4f66"},
}
ADDRESSBOOK = {
    "foo": {"address_key": "bar", "city": "foo", "person": {"first_name": "foo"}},
}

# A rule's 'key' is a path into the resource being looked up, not a column name.
# 'person.first_name' means "the addressbook entry whose person's first_name is
# this", which each backend expresses differently, so the mapping from a logical
# path to a storage predicate belongs here, written out once per backend.
LOOKUPS = {
    ("persons", "first_name"): PERSONS,
    ("addressbook", "person.first_name"): ADDRESSBOOK,
}


class InMemoryResolver:
    """Resolves the lookups the generated rules ask for.

    A real implementation groups the requests the same way and issues one query
    per (kind, key), translating the key into whatever that backend needs: a
    column for a scalar, a JSON path expression in Postgres, a dotted field in
    Mongo, a secondary index in Redis.
    """

    async def resolve(self, requests: list, ctx) -> dict:
        """Resolve every lookup for one request.

        :param list requests: the :class:`RefRequest` objects the rules need
        :param ctx: the read-only request context, for anything a lookup has to be
            scoped by. This example needs none of it; a multi-tenant service would
            take the tenant from here rather than trusting the request body.
        :return: a mapping of request id to the resource found, missing ones left out
        :rtype: dict
        """
        del ctx

        by_lookup = collections.defaultdict(list)
        for request in requests:
            by_lookup[(request.kind, request.key)].append(request)

        found = {}
        for lookup, group in by_lookup.items():
            values = [request.value for request in group]
            _LOGGER.debug(f"Looking up {len(values)} {lookup[0]} by {lookup[1]}: {values}")
            for request in group:
                resource = self.lookup(request)
                if resource is not None:
                    found[request.id] = resource

        return found

    def lookup(self, request: RefRequest):
        """Find a single resource, or None if there is no such thing.

        :param RefRequest request: the kind to look in, the path to match on and
            the value to match
        """
        table = LOOKUPS.get((request.kind, request.key))
        if table is None:
            raise NotImplementedError(
                f"Nothing knows how to look up {request.kind} by {request.key}"
            )

        return table.get(request.value)


def problem_response(error: ValidationError) -> JSONResponse:
    """Render a validation failure as the response the OpenAPI spec advertises.

    RFC 9457 puts the problem at the top level of the body and serves it as
    application/problem+json, so this cannot go through ``HTTPException``: that
    wraps whatever you give it in a ``detail`` key and sends application/json.

    Registering an exception handler for :class:`ValidationError` on the app does
    the same thing for every route at once, which is what you would do for real.
    """
    return JSONResponse(
        status_code=error.status,
        content=error.to_problem(),
        media_type="application/problem+json",
    )

File: examples/addressbook/validation-rs/examples/resolver.rs
//! The one piece of validation you write by hand, in rust.
//!
//! Firestone works out *what* every rule needs; this works out *how* to fetch it.
//! Swap the tables below for your database and nothing else about the generated
//! package changes.
//!
//! Run with `cargo run --example resolver`.

use std::collections::HashMap;

use addressbook_validation::validation::{
    validate, Error, RefRequest, Request, Resolved, Resolver, ResolverError,
};
use async_trait::async_trait;
use serde_json::{json, Value};

/// Stands in for the tables a real service would query.
struct InMemory {
    persons: HashMap<&'static str, Value>,
    addressbook: HashMap<&'static str, Value>,
}

impl InMemory {
    fn new() -> Self {
        InMemory {
            persons: HashMap::from([("foo", json!({"first_name": "foo", "uuid": "3fa8-4f66"}))]),
            addressbook: HashMap::from([("foo", json!({"address_key": "bar", "city": "foo"}))]),
        }
    }

    /// A rule's `key` is a path into the resource being looked up, not a column
    /// name, so the mapping from a logical path to a storage predicate lives here,
    /// written out once per backend.
    fn lookup(&self, request: &RefRequest) -> Option<Value> {
        let value = request.value.as_str()?;
        match (request.kind.as_str(), request.key.as_str()) {
            ("persons", "first_name") => self.persons.get(value).cloned(),
            ("addressbook", "person.first_name") => self.addressbook.get(value).cloned(),
            _ => None,
        }
    }
}

#[async_trait]
impl Resolver for InMemory {
    /// Resolve every lookup for one request.
    ///
    /// A real implementation would group by `(kind, key)` and issue one query per
    /// group. `ctx` is where a multi-tenant service would take the tenant from,
    /// rather than trusting the request body.
    async fn resolve(
        &mut self,
        requests: &[RefRequest],
        _ctx: &Value,
    ) -> Result<Resolved, ResolverError> {
        let mut found = Resolved::new();
        for request in requests {
            if let Some(resource) = self.lookup(request) {
                found.insert(request.id(), resource);
            }
        }

        Ok(found)
    }
}

async fn attempt(label: &str, request: Request<'_>, resolver: &mut InMemory) {
    match validate(request, Some(resolver)).await {
        Ok(()) => println!("{label:<22} accepted"),
        // RFC 9457 puts the problem members at the top level of the body, and it is
        // served as application/problem+json. Your framework's error type probably
        // wraps whatever you hand it, so build the response directly.
        Err(Error::Validation(failed)) => println!(
            "{label:<22} {} {}",
            failed.status(),
            serde_json::to_string(&failed.to_problem()).unwrap()
        ),
        Err(other) => println!("{label:<22} 500 {other}"),
    }
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    let mut resolver = InMemory::new();

    let known = json!({"person": {"first_name": "foo"}, "city": "foo"});
    attempt(
        "known person",
        Request::new("post", "addressbook").body(&known),
        &mut resolver,
    )
    .await;

    let unknown = json!({"person": {"first_name": "nobody"}});
    attempt(
        "unknown person",
        Request::new("post", "addressbook").body(&unknown),
        &mut resolver,
    )
    .await;

    let current = json!({"person": {"first_name": "foo"}, "is_valid": true});
    let flip = json!({"is_valid": false});
    let user = json!({"roles": ["user"]});
    let admin = json!({"roles": ["admin"]});
    attempt(
        "non-admin invalidate",
        Request::new("patch", "addressbook")
            .body(&flip)
            .old(&current)
            .ctx(&user),
        &mut resolver,
    )
    .await;
    attempt(
        "admin invalidate",
        Request::new("patch", "addressbook")
            .body(&flip)
            .old(&current)
            .ctx(&admin),
        &mut resolver,
    )
    .await;

    let in_use = json!({"first_name": "foo"});
    attempt(
        "delete person in use",
        Request::new("delete", "persons").old(&in_use),
        &mut resolver,
    )
    .await;
}
